odl/trafos/backends/pywt_bindings.py

```diff
@@ -0,0 +1,532 @@
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with ODL.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Bindings to the ``PyWavelets`` backend for wavelet transforms."""
+
+# Imports for common Python 2/3 codebase
+from __future__ import print_function, division, absolute_import
+from future import standard_library
+standard_library.install_aliases()
+
+from itertools import product
+import numpy as np
+
+try:
+    import pywt
+    PYWT_AVAILABLE = True
+except ImportError:
+    PYWT_AVAILABLE = False
+
+
+__all__ = ('PAD_MODES_ODL2PYWT', 'PYWT_SUPPORTED_MODES', 'PYWT_AVAILABLE',
+           'pywt_wavelet', 'pywt_mode', 'pywt_coeff_shapes',
+           'pywt_flat_coeff_size',
+           'pywt_flat_array_from_coeffs', 'pywt_coeffs_from_flat_array',
+           'pywt_single_level_decomp', 'pywt_single_level_recon',
+           'pywt_multi_level_decomp', 'pywt_multi_level_recon')
+
+
+PAD_MODES_ODL2PYWT = {'constant': 'zpd',
+                      'periodic': 'ppd',
+                      'symmetric': 'sym',
+                      'order0': 'cpd',
+                      'order1': 'sp1',
+                      'pywt_periodic': 'per'}
+PYWT_SUPPORTED_MODES = PAD_MODES_ODL2PYWT.values()
+
+
+def pywt_wavelet(wavelet):
+    """Convert ``wavelet`` to a ``pywt.Wavelet`` instance."""
+    if isinstance(wavelet, pywt.Wavelet):
+        return wavelet
+    else:
+        return pywt.Wavelet(wavelet)
+
+
+def pywt_mode(pad_mode, pad_const=0):
+    """Convert ODL-style padding mode to pywt-style mode."""
+    pad_mode = str(pad_mode).lower()
+    pad_const = float(pad_const)
+    if pad_mode in PYWT_SUPPORTED_MODES:
+        return pad_mode
+    elif pad_mode == 'constant' and pad_const != 0.0:
+        raise ValueError('constant padding with constant != 0 not supported '
+                         'for `pywt` back-end')
+    else:
+        try:
+            return PAD_MODES_ODL2PYWT[pad_mode]
+        except KeyError:
+            raise ValueError("`pad_mode` '{}' not understood".format(pad_mode))
+
+
+def _check_nlevels(nlevels, data_len, wavelet):
+    """Check if ``nlevels`` is valid for given data length and wavelet."""
+    max_levels = pywt.dwt_max_level(data_len, wavelet.dec_len)
+    if not 0 < nlevels <= max_levels:
+        raise ValueError('`nlevels` must lie between 1 and {}, got {}'
+                         ''.format(max_levels, nlevels))
+
+
+def pywt_coeff_shapes(shape, wavelet, nlevels, mode):
+    """Return a list of coefficient shapes in the specified transform.
+
+    Parameters
+    ----------
+    shape : sequence
+        Shape of an input to the transform.
+    wavelet : string or ``pywt.Wavelet``
+        Specification of the wavelet to be used in the transform.
+        If a string is given, it is converted to a ``pywt.Wavelet``.
+        Use `pywt.wavelist
```

Comment:
I think intersphinx requires a library to export an `objects.inv` file. I'm not sure PyWavelets does that, googling gives no hits. That doesn't have to mean anything, since googling the same for Numpy also only gives the hits on StackOverflow, not anything on the Numpy page itself.